src/linter/xmlTemplate/generator/ControllerByIdDtsGenerator.ts

```diff
@@ -0,0 +1,75 @@
+import ControllerByIdInfo, {IdModulesMap} from "../ControllerByIdInfo.js";
+
+export class ControllerByIdDtsGenerator {
+	// Maps module names to local names
+	private imports = new Map<string, string>();
+
+	generate(controllerByIdInfo: ControllerByIdInfo) {
+		const mappings = controllerByIdInfo.getMappings();
+		if (!mappings.size) {
+			return null;
+		}
+		this.imports = new Map<string, string>();
+		this.addImport("sap/ui/core/mvc/View"); // View is needed for interface ControllerView
+		this.addImport("sap/ui/core/Element"); // Element is needed for byId fallback signature
+		let out = "";
+		mappings.forEach((idToModules, controllerName) => {
+			out += this.generateModuleDeclaration(controllerName, idToModules);
+		});
+		return this.generateCollectedImports() + out;
+	}
+
+	private generateCollectedImports() {
+		let out = "";
+		this.imports.forEach((localName, moduleName) => {
+			out += `import ${localName} from "${moduleName}";\n`;
+		});
+		out += "\n";
+		return out;
+	}
+
+	private generateByIdMapping(idToModules: IdModulesMap) {
+		let out = "\tinterface ByIdMapping {\n";
+		idToModules.forEach((modules, id) => {
+			const localNames: string[] = [];
+			modules.forEach((moduleName) => localNames.push(this.addImport(moduleName)));
+			out += `\t\t"${id}": ${localNames.join(" | ")};\n`;
+		});
+		out += "\t}\n";
+		return out;
+	}
+
+	private generateModuleDeclaration(controllerName: string, idToModules: IdModulesMap) {
+		const moduleName = controllerName.replace(/\./g, "/") + ".controller";
+
+		let out = `declare module "${moduleName}" {\n`;
+		out += this.generateByIdMapping(idToModules);
+		out += `\ttype ByIdFunction = {\n`;
```

Comment:
Template string might be more readable but this is fine too.

Reply:
Thanks. I also didn't like that part and I think in-between refactorings I had reasons for not using a template literal. I took the chance to also go over the other parts with focus on improving readability.